### ISSUE TITLE
Add MANRS certified to Network objects

### ISSUE THREAD

**Issue by mcmanuss8** [peeringdb/peeringdb#801]
Add a field to Networks called "MANRS certified:" with a dropdown with the following values:

- Not certified
- Working to certification
- Certified

Ideally peeringdb would automatically update records based on an API - marking those that are certified as certified and pushing ones that are not that say they are back to Working to certification.

**Comment by job** [comment]
Small comment about the Web UI: I would skip "Not certified" and simply not mention MANRS at all if a network is not part of MANRS. I feel it is good to show a MANRS badge when it is relevant (e.g. the network knew about it and invested some time) but MANRS is not a 'monopolist' in the certification business, so I'd prefer a slightly softer approach.

I'm confident the MANRS people are happy to help automate this through an API or JSON file dump to make this a 100% no-humans-involved process.

**Comment by arnoldnipper** [comment]
How would the MANRS badge then make it into the net's object, @job?

**Comment by arnoldnipper** [comment]
@mcmanuss8, [IXes also may be MANRS compliant](https://www.manrs.org/ixps/ixp-participants/).

**Comment by job** [comment]
> How would the MANRS badge then make it into the net's object, @job?

We could make a small rectangle area where 'badges' can be collected. The moment OpenIX or MANRS indicate that a given ASN or IX is compliant with their requirements, a small icon appears there.

Or we take a tabular approach where rows are added (or removed) with the left column being a short identifier for the badge and the right column a yes/no/status indicator.

In the API JSON output it might be good to make a 'badges' object where as we add more badges over time, more keys appear (with a value of 'not defined', 'yes', 'something'.

**Comment by arnoldnipper** [comment]
Not sure that we need a badges object. Or would we then have additional objects like `ixbadge` and `netbadge`  Something like
```
  "id": 1,
  "badge_id": 1,
  "name":  "MANRS",
  "compliance": "full",
  "ix_id": 31,
  "created": "2020-08-08T00:00:00Z",
  "updated": "2020-08-08T00:00:00Z"
  "status": "ok"
}
```

**Comment by job** [comment]
![Screenshot 2020-08-08 at 8 00 49 PM](https://user-images.githubusercontent.com/153816/89716975-0f296a80-d9b2-11ea-98e2-2a75c561941f.png)

**Comment by shane-kerr** [comment]
Looking at the list of MANRS participants, this seems to me to not be a simple toggle.

Actions 1, 3, 4, and 5 see, binary, but action 2 has sub-actions.

Possibly this should be something like:

> MANRS actions 1: [✓] 2: [⯀] 3: [✗] 4: [✓] 5: [✓]

This would indicate that actions 1, 4, and 5 are implemented, action 3 not, and action 2 partially-implemented. Tooltips could give more detail about what the action is.

**Comment by arnoldnipper** [comment]
> Looking at the list of MANRS participants, this seems to me to not be a simple toggle.

It doesn't have to be. And I would not overload PDB. Why not provide the URL as well e.g. https://www.manrs.org/ixps/ixp-participants/entry/334/

**Comment by shane-kerr** [comment]
> > Looking at the list of MANRS participants, this seems to me to not be a simple toggle.
> 
> It doesn't have to be. And I would not overload PDB. Why not provide the URL as well e.g. https://www.manrs.org/ixps/ixp-participants/entry/334/

I'm a little confused though. So you don't think we need to limit to a simple toggle, but you don't want us to include details in PeeringDB about the MANRS actions? I guess your proposal is a URL instead of a checkbox, and no further details? (I like the URL idea, although if we limit the data to that it means that if anyone is interested in specific MANRS actions they basically have to go to the MANRS site instead of PeeringDB.)

**Comment by arnoldnipper** [comment]
Yes, a checkbox. In case of MANRS it means it is a MANRS participant. But we shouldn't try to explain what MANRS compliance levels are met. Hence the URL.

**Comment by vegu** [comment]
Here's our first revision of specifications - please review and discuss, will move to "Ready to implement" once we're happy

## Abstract

Implement a way to track and determine certification badges on a per entity level, 
with support for all entity types as needed. System should be designed to be open-ended
to allow other badge types easily.

## Considerations

Badges would ideally be automatically updated by querying the provider (MANRS, OpenIX), 
however this might be pending some implementation of the target provider to easily query
for an entity.

In the case of MANRS we have 5 different actions to keep track of. Need to determine
if those are to be grouped in any way, or have different badges for each, or perhaps
require all for a single MANRS badge.

## Specifications

### Database

A single table that keeps track of an entity’s badges by it’s reference tag and id.

- `ref_tag` (ix, net, fac etc)
- `id` (entity id)
- `badge` (MANRS, OIX1, OIX2, etc.)
- `url` (url to certification page, https://www.manrs.org/ixps/ixp-participants/entry/334/)
- `created`
- `updated`

### UI

- Display badges on the right hand side below the other elements as described 
[https://github.com/peeringdb/peeringdb/issues/801#issuecomment-670956810](here)
- Clicking the badge will take you to the certification page

### API

- Add a `badge_set` property to the entities that can have badges

```
“badge_set” : [
    {
	“url”: “https://www.manrs.org/ixps/ixp-participants/entry/334/”
	“badge”: “MANRS”,
	“created”: “2020-08-08T00:00:00Z”
    },
    ...
]
```

- Allow querying by providing a `badge` filter that follows existing api querying functionality
  - `/api/ix?badge=MANRS`
  - `/api/ix?badge__in=MANRS,OIX1`

**Comment by mcmanuss8** [comment]
@peeringdb/pc please review @vegu 's proposal above and vote. 

This sounds good to me, with some small clarifications/questions:
* Do we need to track the id of the other side (ex: a MANRS or Open-IX id?)
* We need a common-ish location in the UI to present the badges for each object so it's moderately consistent across the entity pages
* A cron job (or cron jobs) that run to query from the various APIs and cache them, so we aren't adding real-time dependencies to other systems that aren't mission critical

**Comment by ynbrthr** [comment]
+1

**Comment by martinhannigan** [comment]
How will you manage the hundreds of possible combinations and if they're
legit issued or not? Expired? Revoked? Will we also define BCP vs a real
standard and teach about the different force of law implications? This is
another area of huge potential for fraud and consumer confusion.

I would suggest leaving sales to the sales process and keep PDB out of it.

-1

**Comment by mcmanuss8** [comment]
> How will you manage the hundreds of possible combinations and if they're
legit issued or not? Expired? Revoked?

The idea is to focus strictly on certs that have an API. We'd trust that the certification body would provide a well known API to query (in this case MANRS is queryable via ASN, and ASNs are tightly tied to networks in peeringdb). 

I think there's an argument to be made about wether this adds value to the peering process, but from a "Is this certification legit", I don't think we'd be any better or worse than someone checking MANRS' site directly today.

**Comment by martinhannigan** [comment]
On Wed, Jun 9, 2021 at 19:35 mcmanuss8 ***@***.***> wrote:

> How will you manage the hundreds of possible combinations and if they're
> legit issued or not? Expired? Revoked?
>
> The idea is to focus strictly on certs that have an API. We'd trust that
> the certification body would provide a well known API to query (in this
> case MANRS is queryable via ASN, and ASNs are tightly tied to networks in
> peeringdb).
>
I think there's an argument to be made about wether this adds value to the
> peering process, but from a "Is this certification legit", I don't think
> we'd be any better or worse than someone checking MANRS' site directly
> today.
>

Agree on the latter. If an automation stream is looking for checkboxes it
also seems to not support a strong value proposition. MANRS is also not a
standard. It's a recommendation. That's an important nuance. You can't use
a BCP as a basis for an argument in a court. You can use a bona-fide
standard. If MANRS already has an API seems like a duplication of effort. I
think I'd also add "relevancy" to the list.

-1

**Comment by arobach2** [comment]
> MANRS is also not a standard. It's a recommendation. That's an important nuance. You can't use a BCP as a basis for an argument in a court. You can use a bona-fide standard. If MANRS already has an API seems like a duplication of effort. I think I'd also add "relevancy" to the list.

In fact MANRS is not a BCP, you may look at it as a "soft" certification framework. We do evaluate the applications and check the level of conformance to the best of our abilities. There is also a tool called MANRS Observatory (https://observatory.manrs.org/) that measures "MANRS readiness" of networks. Detailed reports are available to the participants. The MANRS community is also discussing making individual MANRS readiness scores for the MANRS networks available publicly (not just as a tick-boxes in the table (https://www.manrs.org/isps/participants/) 

In this sense the MANRS label signals a certain level of maturity of routing security controls, which maybe useful when making peering policy decisions.

**Comment by martinhannigan** [comment]
First and foremost, there is nothing wrong with MANRS. This discussion is
about letting "things" use PeeringDB as a marketing tool. MANRs is an
example of a thing. In that vein, MANRS is not a bonafide standard. It may
be a de-facto standard. But figuring that out is not something I care
about. Are you saying that it should be a badge on PeeringDB to influence
purchasing decisions? Making PDB a competitive tool is risky business and
going to require much more thought. And potentially money. It's supposed to
be a neutral directory of information useful to facilitate peering. These
types of things take away from that, even if the neutrality impact is by a
third party.

My take away from this discussion is that if a "thing" already has a
mechanism to identify itself and its "users" there is no reason to reinvent
a wheel. Seems reasonable.

YMMV, and warm regards,

-M<

**Comment by arnoldnipper** [comment]
> Good point @martinhannigan - essentially it is duplication of data. If the certification body has an API, then the end user who uses PeeringDB API, can also query MANRS API directly. It is indeed unclear what the value is of copying data from MANRS into PeeringDB and taking up the burden of keeping things well synchronised.

The value is that the end-user saves an extra call. That's also the reason we provide extra data when calling PDB objects (all the read-only fields != DB fields e.g.)

Furthermore, we most likely want to add certification/compliance iff we can get this information via an API call. However, with your argument, we would exactly for that reason not implement it.

**Comment by martinhannigan** [comment]
On Sat, Jul 31, 2021 at 12:07 PM Arnold Nipper ***@***.***>
wrote:

> Furthermore, we most likely want to add certification/compliance iff we
> can get this information via an API call. However, with your argument, we
> would exactly for that reason not implement it.
>
>

Which certifications are you thinking about?

**Comment by arnoldnipper** [comment]
> Which certifications are you thinking about?

None specifically. IMO the issue is back to square one. My conclusion from what @job argues is
- if there is an API for that certification/compliance, the user may use this directly
- if there is no API, then we have to check the certification/compliance manually and most likely won't do it

**Comment by martinhannigan** [comment]
On Sun, Aug 1, 2021 at 5:37 AM Arnold Nipper ***@***.***>
wrote:

> Which certifications are you thinking about?
>
> None specifically. IMO the issue is back to square one. My conclusion from
> what @job <https://github.com/job> argues is
>
>    - if there is an API for that certification/compliance, the user may
>    use this directly
>    - if there is no API, then we have to check the
>    certification/compliance manually and most likely won't do it
>
>
I still can’t think of a reason why if you a user cares they can’t go look
for themselves. These are going to be few and far between and as a result
offer little in benefit compared to the work to implement and maintain.

-1


Warm regards,

-M<

**Comment by arnoldnipper** [comment]
> I still can’t think of a reason why if you a user cares they can’t go look for themselves.

Ease of use? Having all information at your fingertip? That's why we also piggyback additional information when calling PDB objects (like net_count. This information also can easily be derived for those who care)

**Comment by agallo** [comment]
I agree with @arnoldnipper --- easy of use.  It would be nice to see if a potential peer is a MANRS participant.

As a MANRS participant, I would like a place other than the notes field to tell potential peers that we are MANRS compliant.

When looking for potential peers, it would be nice to see this information.  

The current intersection of MANRS participants & peeringDB objects is growing, so while the number may be low today, hopefully in the future, it will not be.

**Comment by therealr2d2** [comment]
We are considering adding some MANRS information in our ISP Portal (Google), as we already are using PeeringDB API it would be relatively easier to query this information there instead of new code to access MANRS API directly only for this info.

It is not super hard to have both, but definitely easier to have all the information in one place.

We agree with Arnold and Andrew on their previous messages and we think that it would add value to have this.

**Comment by job** [comment]
But the information won't be "_in one place_": rather PeeringDB's application would be a sort of caching layer in the middle (which can be out-of-sync). Whether operators query via PeeringDB or via the compliance API directly: new code is needed.

I also think the badges might be subtlety skew how people perceive the interconnection world: while in operational reality MANRS (or any other badges) are not an indicator for safety, reliability, or actually having applied best practices. Many networks out there have configured their routers perfectly but are not MANRS participants.

IMHO: If people are interested in this data, they should fetch it at the source.

-1 from my side

**Comment by martinhannigan** [comment]
Well said. I remain -1. Appreciate everyones input. Thank you.

**Comment by petergitau9** [comment]
I see this as PeeringDB adding value and educating the community. The badges of which MANRS would be one will manage perception positively especially in parts of the world where interconnection changes are infrequent or rarely done and filtering compliance optional unless forced by circumstances. People forget therefore reminding them even if indirectly is a good thing.

**Comment by mcmanuss8** [comment]
I think at this point we either need a *series* of badges we think all add value or none at all. Specifically, we should be considering any of them that would help someone make a decision about peering with a given network, or joining a given ix or facility. 

Can someone propose a series of *other*  that would add similar value? If so, I think we can have a conversation about adding them *all* in, and what value they add to the users of peeringdb. Otherwise, there's enough opposition to doing this just for MANRS that I think we aren't going to.

**Comment by arobach2** [comment]
MANRS is a way for network operators and IXPs to signal their posture on routing security. This signal is useful in the context of peering because it may help inform routing policy decisions, but also stress-test the claim against the interconnection reality - peer pressure. This is why MANRS badge has been implemented in the IXPDB and the IXP Manager. I think exposing it in PeeringDB will add value to the tool and to its users.

I'd also like to note that MANRS is not a self-declared conformance. The secretariat evaluate each application and the MANRS Observatory performs continuous measurement of MANRS readiness. Consider this as a soft certification scheme.

As to other possible badges, I think it makes sense to consider such possibility and implement accordingly, allowing adding them in the future without additional development.

**Comment by therealr2d2** [comment]
>Whether operators query via PeeringDB or via the compliance API directly: new code is needed.

Yes, but less using PeeringDB as it is a marginal addition. To go to the source a full deployment would be required.

I still think that adding the MARNS batch is a good idea as it is helpful to only query 1 API, also, it would be useful to know if the possible peer is MANRS compliance without a few more clicks going to the MANRS website.

And agree with Andrei from the last message.

**Comment by mcmanuss8** [comment]
Hearing no other badges people want and the resistance to just doing this for MANRS, we're going to close this. Can reopen if there's a sufficient list of *other* badges that people would want to see on peeringdb, and at that point we can implement this type of system.
